### PR TITLE
[Relay][MxNet] Add support for foreach

### PR DESCRIPTION
**Pull request by jroesch**
Cherry-pick a small change from Haichen's branch to fix the MxNet converter for Relay's nightly.

@slyubomirsky

### CODE REVIEW

**Comment by slyubomirsky** [comment]
LGTM, thank you Jared

File: python/tvm/relay/frontend/mxnet.py [hunk 1]
```diff
@@ -1096,7 +1256,16 @@ def _from_mxnet_impl(symbol, shape_dict, dtype_info, mod=None):
                 dtype = dtype_info
             node_map[nid] = [_expr.var(node_name, shape=shape, dtype=dtype)]
         elif op_name in _convert_map:
-            res = _convert_map[op_name](children, attrs)
+            # Handle Control-Flow.
+            if op_name in ['_foreach', '_while_loop']:
+                subgraphs = node['subgraphs']
+                assert len(subgraphs) == 1, "only supports loop constructs with a single subgraph"
```
Comment (icemelon9) [inline python/tvm/relay/frontend/mxnet.py]: This is not true for while_loop because it have 2 subgraphs, cond and body.

Comment (jroesch) [reply python/tvm/relay/frontend/mxnet.py]: Yeah I was trying to understand the code, I need to adapt for while loop. I got wrapped up working on paper for ASPLOS, resuming this tomorrow.

File: python/tvm/relay/frontend/mxnet.py [hunk 2]
```diff
@@ -42,6 +47,13 @@
     "relu"   : _op.nn.relu
 }
 
+def _infer_type(node):
```
Comment (zhiics) [inline python/tvm/relay/frontend/mxnet.py]: Looks we have this in `common.py` as well.

File: python/tvm/relay/frontend/mxnet.py [hunk 3]
```diff
@@ -876,6 +888,142 @@ def _lstm_cell(data, states, i2h_weight, h2h_weight, i2h_bias, h2h_bias):
...
+    cond = _mx.symbol.load_json(json.dumps(cond_json))
+    loop_body = _mx.symbol.load_json(json.dumps(loop_body_json))
+
+    cond_input_locs = attrs.get_int_tuple("cond_input_locs")
+    func_input_locs = attrs.get_int_tuple("func_input_locs")
+    func_var_locs = attrs.get_int_tuple("func_var_locs")
+    num_out_data = attrs.get_int("num_out_data")
+    num_outputs = attrs.get_int("num_outputs")
+
+    import pdb; pdb.set_trace()
```
Comment (zhiics) [inline python/tvm/relay/frontend/mxnet.py]: do we need to keep them? same to the below ones.

File: python/tvm/relay/frontend/mxnet.py [hunk 4]
```diff
@@ -876,6 +888,142 @@ def _lstm_cell(data, states, i2h_weight, h2h_weight, i2h_bias, h2h_bias):
...
+
+    return _expr.TupleWrapper(start_loop, 1 + len(prev_states))
+
+
+def _mx_while_loop(inputs, attrs, subgraphs_json, dtype_info, mod):
+    cond_json, loop_body_json = subgraphs_json
+    cond = _mx.symbol.load_json(json.dumps(cond_json))
+    loop_body = _mx.symbol.load_json(json.dumps(loop_body_json))
+
+    cond_input_locs = attrs.get_int_tuple("cond_input_locs")
```
Comment (zhiics) [inline python/tvm/relay/frontend/mxnet.py]: The following several lines seem unused.

File: tests/python/frontend/mxnet/test_graph.py [hunk 5]
```diff
@@ -117,3 +166,5 @@ def relay_compose(F, **kwargs):
     test_dcgan()
     test_squeezenet()
     test_inception_v3()
+    test_foreach_map()
+    test_foreach_fold()
```
Comment (icemelon9) [inline tests/python/frontend/mxnet/test_graph.py]: Could you also add foreach test with both outputs and tests from example 3 in https://mxnet.incubator.apache.org/versions/master/tutorials/control_flow/ControlFlowTutorial.html ?

File: python/tvm/relay/frontend/mxnet.py [hunk 6]
```diff
@@ -1183,8 +1341,10 @@ def from_mxnet(symbol,
         params = {}
         for k, v in symbol.collect_params().items():
             params[k] = _nd.array(v.data().asnumpy())
-        data = mx.sym.Variable("data")
-        sym = symbol(data)
+        inputs = []
```
Comment (icemelon9) [inline python/tvm/relay/frontend/mxnet.py]: These part should be fixed in the master now.

File: python/tvm/relay/frontend/mxnet.py [hunk 7]
```diff
@@ -1105,7 +1260,10 @@ def _from_mxnet_impl(symbol, shape_dict, dtype_info, mod=None):
             raise tvm.error.OpNotImplemented(
                 'Operator {} is not supported in frontend MXNet.'.format(op_name))
 
-    outputs = [node_map[e[0]][e[1]] for e in jgraph["heads"]]
+    outputs = []
```
Comment (icemelon9) [inline python/tvm/relay/frontend/mxnet.py]: seems these lines should be equivalent to original code

File: python/tvm/relay/loops.py [hunk 8]
```diff
@@ -63,3 +69,53 @@ def while_loop(cond, loop_vars, loop_bodies):
...
+    state_vars = []
+    for i, st in enumerate(init_states):
+        state_ty = sb.type_of(st)
+        state_vars.append(
+            _expr.var(f"st{i}", type_annotation=state_ty))
+
+    data_ty = sb.type_of(tensor)
+    data_var = _expr.var("data", type_annotation=data_ty)
+
+    if axis is None:
```
Comment (icemelon9) [inline python/tvm/relay/loops.py]: should we just use `axis=0` in the function?

File: python/tvm/relay/frontend/mxnet.py (hunk 7 above)
Comment (jroesch) [reply python/tvm/relay/frontend/mxnet.py]: This is for debugging purposes, the original code is hard to insert debugging.

File: python/tvm/relay/frontend/mxnet.py (hunk 6 above)
Comment (jroesch) [reply python/tvm/relay/frontend/mxnet.py]: I'll rebase.

File: python/tvm/relay/loops.py [hunk 9]
```diff
@@ -63,3 +69,53 @@ def while_loop(cond, loop_vars, loop_bodies):
     func = _expr.Function(fresh_vars, sb.get())
     let = _expr.Let(loop, func, loop)
     return let
+
+def foreach(tensor, iter, init_states, axis=None, loop_name=None, mod=None):
+    assert isinstance(tensor, _expr.Expr), "data to slice must be a tensor"
+    assert isinstance(init_states, list), "initial states must be a list"
+
+    Prelude(mod)
```
[the remark this answers is not in the document]
Comment (jroesch) [reply python/tvm/relay/loops.py]: Yeah I don't really love how the Prelude works right now ... this is an issue we should talk about.

[the remark this answers is not in the document]
Comment (jroesch) [reply python/tvm/relay/loops.py]: @MarisaKirisame

File: python/tvm/relay/frontend/mxnet.py (hunk 3 above)
Comment (jroesch) [reply python/tvm/relay/frontend/mxnet.py]: No, this was corrected in local commit will follow up

**Comment by slyubomirsky** [comment]
@jroesch Could we try to revive this PR since our benchmarks use it?

**Pull request closed without merging**